### PR TITLE
feat(Locomotion): translate play area jump into game jump

### PR DESCRIPTION
**Pull request by ajaxlex**
Adds the player jump script, and demo scene 039 to exercise it.

### CODE REVIEW

**Comment by ajaxlex** [comment]
Sometimes you have to nuke it from orbit, just to be sure.

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 1]
```diff
@@ -0,0 +1,366 @@
+// Player Jump|Scripts|0240
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: I'd put this next to Player climb (which is 0210) so maybe make this (0215)

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+{
+    using UnityEngine;
+    using System.IO;
+    using System;
+
+    /// <summary>
+    /// The player jump provides a jump behavior by tracking HMD world position over time, calculating velocities of a crouch and a spring motion, and adding velocity to the attached rigidbody on playerpresence.  It should be attached to the `[CameraRig]` object.
+    /// </summary>
+    /// <remarks>
+    /// This relies heavily on the status of the playerPresence, and will probably have to be updated as that changes.  In particular, playerPresence sometimes deactivates physics, which can leave this in a stuck state
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: I wouldn't mention that things may change. Write the docs as if this is the way it's always going to work. If things change, we'll update the docs to reflect the changes

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    /// the checkForJumpEnd method sends a startPhysicsFall with a velocity of 0 to prevent this issue ( cf. HeightAdjustTeleport )
+    /// 
+    /// Jump scaling for non-fixed jump factors is calculated in CalcAdjustedFinalJumpVelocity
+    /// </remarks>
+    /// <example>
+    /// VRTK/Examples/039_CameraRig_PlayerJump
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: Look at another script for the correct sentence to use when mentioning an example.

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        private int historyLen = 5;
+
+        private Transform HMDTransform;
+        private VRTK_PlayerPresence playerPresence;
+
+        #endregion
+
+        #region ### private logging vars ###
+
+        // set to writable directory eg. C:/Users/Public/Documents/Unity Projects/MyProject/Assets
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: I'm not sure including custom logging is a good idea.

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+        #endregion
+
+        private void Awake()
+        {
+            // Required Component: VRTK_PlayerPresence
+            playerPresence = GetComponent<VRTK_PlayerPresence>();
+            if (!playerPresence)
+            {
+                playerPresence = gameObject.AddComponent<VRTK_PlayerPresence>();
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: If a PlayerPresence is always required then use the [RequiredComponent] unity annotation instead so less magic happens at runtime

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+                if (CrouchWasDetected())
+                {
+                    UpdateCrouchLowpoint();
+
+                    if (CrouchTimerNotExpired())
+                    {
+                        if (VelocityChangeAboveThreshold(dv))
+                        {
+                            // if head position exceeds the baseline
+                            if (HMDHeightExceedsBaseline())
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: can these nested if's be squished into one conditional statement to remove some of the nesting?

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+                                DoJump(jumpVelocity);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        crouchStartTime = 0;
+                        crouchLowTime = Time.time;
+                        crouchLow = baselineHeight + 1000000; // TODO probably there is a better way
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: worth replacing this number with a variable with a meaningful name so you can understand what the 100000 is for.

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+                        jumping = false;
+                    }
+                }
+                else
+                {
+                    jumping = false;
+                }
+            }
+            if (debugging)
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: take this out, as the content is just commented out anyway. If you want to keep a dev branch with these test ideas in then that's fine but shouldn't be commited to master

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+            float lowpointElapsed = Time.time - crouchLowTime;
+
+            // it may be that the time component should be deemphasized to exaggerated the height change vs the time delay
+            float velocity = lowpointElapsed <= 0 ? 0 : (baselineHeight - crouchLow) / lowpointElapsed;
+
+            // This calc attempts a nicer scaling per player effort.
+            // could possibly be interpolated between the up and down threshold values or something to make it have some sense
+            float scaled = velocity * jumpScaling;
+
+            if (debugging)
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: remove unused code

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+        }
+
+        private void UpdateCrouchLowpoint()
+        {
+            if (HMDTransform.position.y < crouchLow)
+            {
+                crouchLow = HMDTransform.position.y;
+                crouchLowTime = Time.time;
+
+                //DebugHelper(" Update CrouchLow: " + crouchLow + " CLT: " + crouchLowTime);
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: remove

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+            }
+        }
+
+        private void CheckForJumpEnd(float dv)
+        {
+            if (Mathf.Abs(dv) < baselineUpdateSensitivity)
+            {
+                jumping = false;
+                crouchStartTime = 0;
+                //DebugHelper("Ending jump - isFalling = false");
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: remove

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+        private void DoJump(float jumpVelocity)
+        {
+            Vector3 forward = GetPlayerForward() * jumpVelocity;
+
+            Vector3 calcVelocity;
+            calcVelocity.x = forwardFactor * forward.x;
+            calcVelocity.y = upFactor * forward.y;
+            calcVelocity.z = forwardFactor * forward.z;
+
+            //DebugHelper("JUMP EVENT: " + calcVelocity.ToString());
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: remove

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+
+            playerPresence.StartPhysicsFall(Vector3.zero);
+
+            playerPresence.GetComponent<Rigidbody>().velocity = calcVelocity;
+        }
+
+        private Vector3 GetPlayerForward()
+        {
+            // necessaries for triangulation
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: again, remove this commented out code and just keep things like this in your own dev branch

File: Assets/VRTK/Scripts/VRTK_PlayerJump.cs [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+            Vector3 gaze = HMDTransform.forward;
+            return new Vector3(gaze.x, 1, gaze.z);
+        }
+
+        private void OnApplicationQuit()
+        {
+            QuitLogging();
+        }
+
+        #region ### Private Logging and Debugging Methods ###
```
Comment (thestonefox) [inline Assets/VRTK/Scripts/VRTK_PlayerJump.cs]: I think it's probably best to remove custom logging, it's only going to confuse someone down the line :)

**Comment by ajaxlex** [comment]
SO IT IS ASKED.....SO IS IT DONE!
( i think ).  Also, do you prefer amends only for these feature branches, or could I just have made a separate commit?

**Comment by thestonefox** [comment]
all the code looks good, i'll check out the PR a bit later and give it a test to give it the final check

**Comment by thestonefox** [comment]
Just realised, you need to make sure your master branch is up to date.

There is already a example scene 39. i think the next available one is 41 now

**Comment by thestonefox** [comment]
I'm also struggling to get this consistently working. may be worth getting some others to check out the PR for some feedback. I'll post in the slack channel

**Comment by thestonefox** [comment]
did you forget to do a rebase?

you have a "lot" of extra commits :(

**Comment by ajaxlex** [comment]
No, I rebased.  These are whitespace errs, I think.  I asked for a rebase --whitespace=fix

I am going to have to scale back my ambitions for open source contribution at this rate!  I may just start again.  It is far easier to copy the code in than to f-around with all the minute conflicts.

**Comment by ajaxlex** [comment]
I guess I am mistaken.  Whatever.  Time to get mopping

**Comment by thestonefox** [comment]
Sometimes it is easier to just start on a fresh branch and copy the changes over. I do that all the time :)

**Pull request closed without merging**